WIP: Feat/support keystone

ZhenQian-keystone opened the pull request:
1、add keystone connect wallet support

2、add keystone sign tx support

Perronef5 commented on src/App.tsx:
```diff
@@ -124,46 +125,50 @@ const App = () => {
-                                            <NetworkAwareStatusBar />
-                                            <RootNavigator />
+                      <KeystoneOnboardingProvider
+                        baseUrl={Config.ONBOARDING_API_URL}
```
What is this base url for?

ZhenQian-keystone replied:
remove unused code , thanks

Perronef5 commented on src/assets/images/connectKeystoneLogo.png:
Can we get @2x and @3x version of the png please

ZhenQian-keystone replied:
resolved

Perronef5 commented on src/components/CameraScannerLayout.tsx:
```diff
@@ -0,0 +1,101 @@
+        <Box
+          position="absolute"
+          borderColor="hntBlue"
+          width={40}
```
Should we also have a constant for 40?

ZhenQian-keystone replied:
resolved add BORDER_SEGMENT_SIZE constant for 40

Perronef5 commented on src/components/DaynamicQrScanner.tsx:
```diff
@@ -0,0 +1,96 @@
+  if (!hasPermission) {
+    return (
+      <BackScreen>
+        <Box />
```
Can we display some error to the user vs a blank screen?

ZhenQian-keystone replied:
resolved

Perronef5 commented on src/features/browser/BrowserWebViewScreen.tsx:
```diff
@@ -270,39 +270,22 @@ const BrowserWebViewScreen = () => {
+            const signedTransaction =
+              await anchorProvider?.wallet.signTransaction(convertTx)
+            signedTransactions.push(signedTransaction)
+            await new Promise((resolve) => setTimeout(resolve, 2000))
```
Why are we waiting 2 seconds?

ZhenQian-keystone replied:
This issue arises when using Jupiter for swap transactions. Jupiter constructs two transactions, which means the cold wallet needs to sign twice. In terms of UI interaction, we need to open the modal twice in succession for QR code scanning and signing. However, a strange thing happens: while the first bottomSheetModalRef.current?.present() call successfully brings up the modal, the second call after completion fails to wake up the modal. So I added a 2-second sleep, after which the second modal was successfully awakened. I'm confused - is this caused by React Native performance issues or something else? I haven't been able to find the real reason for this problem. Can you help me?

Perronef5 commented on src/features/keystone/KeystoneOnboardingProvider.tsx:
```diff
@@ -0,0 +1,87 @@
+  baseUrl,
+}: {
+  children: ReactNode
+  baseUrl: string
```
Is this being used?

ZhenQian-keystone replied:
I double-checked and removed the unused code. Thanks.

Perronef5 commented on src/features/keystone/SelectKeystoneAccountsScreen.tsx:
```diff
@@ -0,0 +1,259 @@
+          Select Keystone Accounts
+        </Text>
+        <Text textAlign="center" p="s" variant="body1" mb="l">
+          A secret phrase can be used to generate multiple wallets by using
```
Add to translations please

ZhenQian-keystone replied:
resolved

Perronef5 commented on src/features/keystone/SelectKeystoneAccountsScreen.tsx:
```diff
@@ -0,0 +1,259 @@
+          lineHeight={44}
+          mb="s"
+        >
+          Select Keystone Accounts
```
Add to translations too

ZhenQian-keystone replied:
resolved

Perronef5 commented on package.json:
```diff
@@ -57,13 +57,15 @@
     "@maplibre/maplibre-react-native": "^10.0.0-alpha.2",
     "@metaplex-foundation/mpl-bubblegum": "0.6.0",
     "@metaplex-foundation/mpl-token-metadata": "2.10.0",
+    "@ngraveio/bc-ur": "^1.1.13",
```
What does this add?

ZhenQian-keystone replied:
Because Keystone follows the protocol specified in https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-005-ur.md and uses animated QR codes for data transmission, in this process, structured data needs to be first encoded and decoded using the UR protocol. Therefore, it is necessary to introduce the bc-ur library to decode the data.

https://github.com/ngraveio/bc-ur

Perronef5 reviewed (commented):
Overall PR is looking solid just a few comments here and there. Can we get a demo of how this all looks for an end to end flow. Thanks!

ZhenQian-keystone commented:
> Overall PR is looking solid just a few comments here and there. Can we get a demo of how this all looks for an end to end flow. Thanks!

Thank you for your reply. I will make the modifications one by one. We will record a video and send it to the Telegram group.

ZhenQian-keystone commented on src/components/DaynamicQrScanner.tsx:
```diff
@@ -0,0 +1,96 @@
+  progress: number
+  onBarCodeScanned: (data: string) => void
+}
+const DaynamicQrScanner = ({ onBarCodeScanned, progress }: Props) => {
```
resolved

ZhenQian-keystone commented on src/features/keystone/KeystoneOnboardingProvider.tsx:
```diff
@@ -0,0 +1,87 @@
+const useKeystoneOnboardingHook = ({ baseUrl }: { baseUrl: string }) => {
+  const initialState = useMemo(() => {
+    return {
+      netType: NetType.MAINNET,
```
resolved  The network type may not be necessary because Keystone will derive the corresponding path's public key on the cold wallet side.

ChewingGlass commented on src/features/keystone/KeystoneAccountAssignScreen.tsx:
```diff
@@ -0,0 +1,212 @@
+    const accountBulk = keystoneOnboardingData.accounts.map(
+      (account, index) => ({
+        alias: getName(index),
+        address: base58.encode(
```
I don't think that this will work. `address` is expected to be a Helium L1 style address (gross, I know, we mostly use `solanaAddress` now). 

https://github.com/helium/helium-program-library/blob/master/packages/migration-service/src/index.ts#L75-L91

This code might be helpful as a reference.

ZhenQian-keystone replied:
Because the Keystone cold wallet only exports the Solana address in its data, I assigned the address value to be the Solana public key.

ChewingGlass replied:
Helium's old L1 used ED25119 keys. That means for every solana pubkey, there's a valid helium l1 representation of that key. That's what the code I linked above does. You should not put the solana address here. You should convert that address to a helium address. Unfortunately this is just some tech debt in this code since it used to run against the helium l1

ZhenQian-keystone replied:
resolved

ChewingGlass commented on src/storage/cloudStorage.ts:
```diff
@@ -63,18 +72,28 @@ export const sortAccounts = (
+  const sortedByAliasWithSolanaAddress = sortedByAlias.map((acct) => {
+    // here we need to check if the account is keystone account
+    if (acct.keystoneDevice) {
+      return { ...acct, solanaAddress: acct.address }
```
Why does this need to happen?

ZhenQian-keystone replied:
I encountered a very strange problem. When using the Keystone cold wallet to import a Solana account, the data was stored completely, meaning the solanaAddress had a value. However, when the onboarding process ended and I tried to read the data on the account interface, I found that the solanaAddress had somehow become an empty string ("") for an unknown reason. To solve this strange issue, I intercepted the data before it was loaded from storage into the provider, and manually assigned the address value. This way, it could be displayed normally.

ZhenQian-keystone replied:
What do you think might be the reason for this phenomenon? I'm looking forward to your response.

ChewingGlass replied:
It may be that you aren't properly writing a helium version of the solana address in the account? So some code later that write solanaAddress as a derivation of the `address` field writes empty?

ZhenQian-keystone replied:
i am a bit confused. Could you tell me where the logic for changing the solanaAddress to an empty string is? Can you provide a link to it? I haven't been able to find it. I'd really appreciate it if you could.

ChewingGlass replied:
See the above comment on not setting the proper helium `address`

ZhenQian-keystone replied:
I wrote a demo to convert a Solana wallet public key to a Helium old address. Is this conversion logic correct?

```
// convet solana public key to helium wallet address and back
const { PublicKey } = require("@solana/web3.js");
const { ED25519_KEY_TYPE } = require("@helium/address/build/KeyTypes");
const  Address = require("@helium/address").default;

// solana wallet path
// "m/44'/501'/0'"
// solana wallet public key
// "2b8d8b3addd92759f55b840c2852f5bd50aee3552fe987ee0d4fe24b9043df8e"
// solana wallet address
// 3w1iMvjKGxpbGaaSekNUsZBcVKERg2BCsUZMGrjcTMsj
let solWalletPublicKey = Buffer.from("2b8d8b3addd92759f55b840c2852f5bd50aee3552fe987ee0d4fe24b9043df8e", "hex");
let solWalletAddress = new PublicKey(solWalletPublicKey).toBase58();
console.log(solWalletAddress) // 3w1iMvjKGxpbGaaSekNUsZBcVKERg2BCsUZMGrjcTMsj


// convert solana public key to helium wallet address
const pkey = new PublicKey(solWalletPublicKey);
const heliumAddr = new Address(0, 0, ED25519_KEY_TYPE, pkey.toBytes());
let heliumAddress = heliumAddr.b58;
console.log(heliumAddress); // 13GvgGD1zwkBw5ZihDvXz81iMiJG97o3xL5Sz1wW2ffUUNHySxX


// convert helium wallet address to solana public key
const addr = Address.fromB58(heliumAddress);
let solanaAddress = new PublicKey(addr.publicKey).toBase58();
console.log(solanaAddress); // 3w1iMvjKGxpbGaaSekNUsZBcVKERg2BCsUZMGrjcTMsj
```

If this logic is correct, then I will modify the logic in the above code. In other words, when I get the Solana public key from the keystone cold wallet, I need to first convert it to a Helium old address and assign it to the 'address' variable. The 'solanaAddress' will be automatically assigned after completing the migrate operation.

ZhenQian-keystone replied:
So, about what you said regarding some code that resets the solanaAddress to empty - I've been trying to find this piece of logic. Can you tell me where it is? I've been looking all over for it but can't seem to locate it.

Perronef5 reviewed (changes requested):
Tested this locally and saw a couple issues:

1. When you scan a QR code that is not expected when connecting your keystone the app crashes. We need to handle this crash. (Happy path works as expected.)
2. Unable to remove keystone wallet from within the settings.

Once these are fixed happy to test again 🙂

ZhenQian-keystone commented:
> Tested this locally and saw a couple issues:
> 
> 1. When you scan a QR code that is not expected when connecting your keystone the app crashes. We need to handle this crash. (Happy path works as expected.)
> 2. Unable to remove keystone wallet from within the settings.
> 
> Once these are fixed happy to test again 🙂

1、I fixed the issue where the app would crash when scanning an unknown QR code. I also added an alert to give users a heads up.

2、I tried to reproduce the problem with removing wallets on iOS devices, but it's working fine for me. Could you give me more details about what's going wrong? I'm having trouble seeing the issue on my end.

Perronef5 reviewed (commented):
Tested your most recent changes and working well. Branch needs a rebase then I can approve 👍

ZhenQian-keystone commented:
> Tested your most recent changes and working well. Branch needs a rebase then I can approve 👍

done
